test/unit/offset.js

```diff
@@ -568,4 +568,142 @@ QUnit.test( "iframe scrollTop/Left (see gh-1945)", function( assert ) {
 	}
 } );
 
+(function() {
+	var POSITION_VALUES = [ "static", "relative", "absolute", "fixed" ];
+
+	supportjQuery.each( POSITION_VALUES, function( i, docPosition ) {
+		supportjQuery.each( POSITION_VALUES, function( i, bodyPosition ) {
+			supportjQuery.each( [ "static", "absolute" ], function( i, markerPosition ) {
+
+				testIframe( "coordinates relative to document" +
+							" (html{position:" + docPosition + "} body{position:" + bodyPosition + "} div{position:" + markerPosition + "})",
+						"offset/rel-doc.html", function( assert, $, iframe, doc ) {
+					assert.expect( 256 );
+
+					var docElem = doc.documentElement, // `<html>` element
+						bodyElem = doc.body,
+						marker1 = doc.getElementById( "marker1" ),
+						marker2 = doc.getElementById( "marker2" ),
+						$marker1 = $( marker1 ),
+						$marker2 = $( marker2 ),
+						affectProps, markerTopLeft,
+
+						// Bit as Flag of box-properties
+						FLAG_DOC_MARGIN = 1,
+						FLAG_DOC_BORDER = 2,
+						FLAG_DOC_PADDING = 4,
+						FLAG_BODY_MARGIN = 8,
+						FLAG_BODY_BORDER = 16,
+						FLAG_BODY_PADDING = 32,
```

Comment:
Sorry, I have to learn English...
So, please check following:

- I interpreted [your comment in that issue](https://github.com/jquery/jquery/issues/3080#issuecomment-215738621) as that the code should be tested with changing `margin` and `border`. Did I mistake again?

- All `assert.deepEqual()`s that are passed  return value from `.offset()` failed.
I found this:
https://github.com/jquery/jquery/blob/master/external/qunit/qunit.js#L1628
`if ( a.constructor === b.constructor ) {`
Expected values that were made in the test code and return values from `.offset()` that were made in `iframe` have constructor in different namespaces (i.e. window), therefore `a.constructor === b.constructor` is `false`.
Return values from `.offset()` have to be copied to be checked by `assert.deepEqual()`.
`supportjQuery.extend({}, $( "#static" ).offset())`

- This test failed when `html{position}` is `static` and `body{position}` is `non-static`.
```js
assert.deepEqual( absoluteOffset, { top: origin + 20, left: origin + 20 },
    "offset of position:absolute element ignores box styles of position:static ancestors" );
```

For example, in a case `html{position:static} body{position:relative}`:
```
Expected: 

{
  "left": 300,
  "top": 300
}

Result: 

{
  "left": 330,
  "top": 330
}
```

I think, this code is incorrect:

```js
var origin = 0 +
    (docPosition === "static" ? 0 : 10 + 20) +
    (bodyPosition === "static" ? 0 : 40 + 80 + 160);
```

If `body` has `position:non-static`, the element is positioned relative to `body` without being affected by `position` of `<html>`. `body` was already positioned with `margin`, `border` and `padding` of `<html>`.
Therefore the code should be:

```js
var origin =
    bodyPosition !== "static" ? 10 + 20 + 40 + 80 + 160 : // `10 + 20 + 40` must be included
    docPosition !== "static" ? 10 + 20 : // relative to inside `<html>`
    0; // relative to document (i.e. outside `<html>`)
```

- This test failed when `html{position}` is not `static` or `relative`.
```js
assert.deepEqual(
    supportjQuery.extend({}, $( "#static" ).offset()),
    { top: bodyContentOrigin + originAdjust, left: bodyContentOrigin + originAdjust },
    "offset of position:static element respects ancestor positioning" );
```

I think, this code is incorrect:

```js
if ( docPosition !== "static" ) {
    originAdjust += 15 - (docPosition === "relative" ? 0 : 10);
}
```

If `<html>` has `position:non-static`, it is positioned with `margin` even if `potision` is other than `relative`.
Therefore the code should be:

```js
if ( docPosition !== "static" ) {
    originAdjust += 15;
}
```

- This test failed in some cases:

```js
assert.deepEqual(
    absoluteOffset,
    { top: origin + originAdjust + 20, left: origin + originAdjust + 20 },
    "offset of position:absolute respects ancestor positioning" );
```

This is solved by fixing `originAdjust` and `origin`.